### ISSUE TITLE
Session no longer exists errors need to be handled

### ISSUE THREAD

**Issue by highlyunavailable** [PlayFab/consuldotnet#7]
From @rmul:

I have an issue and would like to check if you think this is related:
When a semaphore has been aqcuired and the session gets lost (for instance use the api to destroy the session) the RenewPeriodic method throws a "Session no longer exists" exception but catches that also and a cycle later a "Session timed out" exception is thrown which is not handled by the semaphore.acquire method.
If it is not related I am more than willing to create a new issue and help out with it.

---

This is a broader error in that both locks and semaphores need to be looked at.

**Comment by highlyunavailable** [comment]
@rmul This has been resolved, you should be able to get the new version from https://www.nuget.org/packages/Consul/0.5.2.5 or from this repo.

**Comment by rmul** [comment]
@highlyunavailable I've tested the new version, but now it seems that the sessionrenew is not run; The semaphore gets lost every sessionTTl (10 secs) and then gets re-aquired..

**Comment by highlyunavailable** [comment]
@rmul I'm not seeing that behavior, actually.

I added a test method that looks like this (I grabbed some of the bits from your code, just to see if it was a weird limit or something causing it):

```csharp
[TestMethod]
public void Semaphore_AcquireWaitRelease()
{
    const string Name = "XService";
    const string Prefix = Name + "/lock";
    const int Limit = 1;
    var _semaphoreOptions = new SemaphoreOptions(Prefix, Limit) { SessionName = Name + "_Session", SessionTTL = TimeSpan.FromSeconds(10) };
    var c = ClientTest.MakeClient();

    var s = c.Semaphore("test/semaphore", 2);

    s.Acquire(CancellationToken.None);

    Assert.IsTrue(s.IsHeld);
    Thread.Sleep(30000);
    Assert.IsTrue(s.IsHeld);

    Assert.IsTrue(s.IsHeld);

    s.Release();

    Assert.IsFalse(s.IsHeld);
}
```

with the following Consul output:

```
    2015/05/26 06:42:37 [DEBUG] http: Request /v1/session/create (2.0014ms)
    2015/05/26 06:42:37 [DEBUG] http: Request /v1/kv/test/semaphore/61b1fd08-652c-b98e-2302-b70d886b9ade?flags=16210313421097356768&acquire=61b1fd08-652c-b98e-2302-b70d886b9ade (2.0011ms)
    2015/05/26 06:42:37 [DEBUG] http: Request /v1/kv/test/semaphore?recurse&wait=15s (0)
    2015/05/26 06:42:37 [DEBUG] http: Request /v1/kv/test/semaphore/.lock?flags=16210313421097356768&cas=10 (999.1µs)
    2015/05/26 06:42:37 [DEBUG] http: Request /v1/kv/test/semaphore?recurse&consistent (0)
    2015/05/26 06:42:45 [DEBUG] http: Request /v1/session/renew/61b1fd08-652c-b98e-2302-b70d886b9ade (0)
    2015/05/26 06:42:52 [DEBUG] http: Request /v1/session/renew/61b1fd08-652c-b98e-2302-b70d886b9ade (0)
    2015/05/26 06:43:07 [DEBUG] http: Request /v1/kv/test/semaphore/.lock (0)
    2015/05/26 06:43:07 [DEBUG] http: Request /v1/kv/test/semaphore/.lock?flags=16210313421097356768&cas=19 (2ms)
    2015/05/26 06:43:07 [DEBUG] http: Request /v1/kv/test/semaphore?recurse&consistent&index=19 (30.003102s)
    2015/05/26 06:43:07 [DEBUG] http: Request /v1/kv/test/semaphore/61b1fd08-652c-b98e-2302-b70d886b9ade (1.0006ms)
    2015/05/26 06:43:22 [DEBUG] consul.state: Session 61b1fd08-652c-b98e-2302-b70d886b9ade TTL expired
    2015/05/26 06:43:22 [DEBUG] consul.state: Invalidating session 61b1fd08-652c-b98e-2302-b70d886b9ade due to session destroy
```

It's calling Renew twice in there, which to me says it's working.

**Comment by rmul** [comment]
Ok, I'll check my code and see if I can still reproduce

**Comment by highlyunavailable** [comment]
@rmul It looks like you're using an older version. I cloned a copy of your project and swapped out the consul DLL files with one from my build and it appears to be renewing sessions.

**Comment by highlyunavailable** [comment]
Fixed by 65b819d2fc828cc6b7455acc18a330ebd6278d2b - thanks @rmul 

Added tests to verify this is fixed for locks as well in ef8b00dd55c9fe87d2eefceeb0a58b06b8531436
